### PR TITLE
Initial implementation of strict-containers

### PR DESCRIPTION
**Pull request by infinity0**
As per #22, it turns out that duplicating the code then applying a few small patches, is easier than maintaining a mirror API via newtype and manually adding ! annotations on all the mirror API functions.

There is a minor "perfectness" issue with HashMap that I will open an issues on their side to discuss.

### CODE REVIEW

File: strict-containers/patches/Sequence.patch [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ 
+ data Digit a
+-    = One a
+-    | Two a a
+-    | Three a a a
+-    | Four a a a a
++    = One !a
++    | Two !a !a
++    | Three !a !a !a
++    | Four !a !a !a !a
```
Comment (phadej) [inline strict-containers/patches/Sequence.patch:25]: I remember that `Sequence` relies on lazyness to provide (amortized) runtime complexities. I think this change just throws them out of the window - you have to show them new complexities and update the docs.

File: strict-containers/src/Data/Strict/ContainersUtils/Autogen/StrictPair.hs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{-# LANGUAGE CPP #-}
+#if !defined(TESTING) && defined(__GLASGOW_HASKELL__)
+{-# LANGUAGE Safe #-}
+#endif
+
+#include "containers.h"
+
+-- | A strict pair
+
+module Data.Strict.ContainersUtils.Autogen.StrictPair (StrictPair(..), toPair) where
```
Comment (phadej) [inline strict-containers/src/Data/Strict/ContainersUtils/Autogen/StrictPair.hs:10]: unfortunate duplication of what is already in `strict`

**Review by phadej — COMMENTED** [review]
I'm not sure I like this experiment living in `strict` repository. Does it have to?

Concretely, it complicates the low-maintenance of `strict` stuff which is essentially done.

File: strict-containers/patches/Sequence.patch (hunk 1 above)
Comment (infinity0) [reply strict-containers/patches/Sequence.patch:25]: As I understand, the laziness performance gain is on parts of the data structure itself, not of the elements. Specifically, it is via this part, which remains unchanged in this PR:

```haskell
data FingerTree a
  {- ... -}
    | Deep {-# UNPACK #-} !Int !(Digit a) (FingerTree (Node a)) !(Digit a)  -- lazy subtree
```

File: strict-containers/src/Data/Strict/ContainersUtils/Autogen/StrictPair.hs (hunk 2 above)
Comment (infinity0) [reply strict-containers/src/Data/Strict/ContainersUtils/Autogen/StrictPair.hs:10]: Yes, we could file a PR to them to use `strict`, then re-running `regen.sh` will automatically pull this in and drop it here too.

**Comment by infinity0** [comment]
Sure, I can move this to a separate repo.

**Comment by phadej** [comment]
Separate repository (outside of `haskellari`) would be nice. I won't need to worry about it then.

**Comment by infinity0** [comment]
I'd like to keep it in haskellari so it's easier to have co-maintainers if that's OK? I don't want to solo maintain this.

**Comment by phadej** [comment]
Sorry, I have to say no to  haskellari. I don't want to give any hints that I'm affiliated, because I wouldn't.

`strict-containers` have to prove its utility and stability first, currently it doesn't even exist.

File: strict-containers/patches/Sequence.patch (hunk 1 above)
Comment (infinity0) [reply strict-containers/patches/Sequence.patch:25]: From the paper: "In a strict language that provides a lazy evaluation primitive, we need only suspend the middle subtree of each Deep node, so only Θ(log n) suspensions are required in a tree of size n. Even in a lazy language, some space could be saved in practice by ensuring that these were the only suspensions in the tree, for example by using Haskell’s strictness annotations." i.e. making everything else strict preserves the performance bounds

**Comment by infinity0** [comment]
That's quite disappointing and unexpected when I signed up to be a part of this. What do you consider the barrier to be to "prove its utility and stability"?

**Comment by infinity0** [comment]
One reason I'd like to maintain this together with strict, is so we can have a unified approach to where instances go, and to that end I in fact would welcome your (and others') opinions on this matter. In this PR I did not yet define instances for binary, serialise, lens or optics, and that's an open question.

**Comment by phadej** [comment]
Please let us do not mix `strict` and `strict-containers` issues.

I was a maintainer of `strict` and `strict-base-types` and welcomed an effort to cleanup those and push its usage forward (to `aeson`, `lens` etc.). This was stuff I have thought about, and you offered a very needed push. I'm thankful for that.

On the other hand, `strict-containers` is new thing, it poses a lot of new questions, some of which you yourself bring up. That is all something I would pass on. (The instance question is hard, e.g. https://hackage.haskell.org/package/ordered-containers simply doesn't have any instances, https://hackage.haskell.org/package/insert-ordered-containers depends on both `lens` and `optics` as the only package using it depends on both itself).

I'm sorry, but I don't have interest and bandwidth for `strict-containers`. I care about `strict`; about `strict-containers` I don't.

**Comment by infinity0** [comment]
OK, I won't expect you to help co-maintain it but can we keep it as a separate repo in the organisation. When I first signed up we agreed that it would not just be "Oleg's stuff". `strict-containers` should be feature-complete from essentially the initial version, the only purpose really is to mirror `containers` and `unordered-containers` and the API is supposed to basically be the same. The question of instances is a general Haskell issue and not specific to `strict-containers` either, so it is not something that should block where this repo goes.

**Comment by phadej** [comment]
Ok. Go for a new repository in `haskellari` then.

**Pull request closed without merging**